### PR TITLE
Vb 1921 user information from token and save on database

### PR DESCRIPTION
**Pull request by dhirajshettymoj**
## What does this pull request do?

Saves actionedBy to the database. 

## What is the intent behind these changes?

Auditing and to retrieve audit information for a visit.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt [hunk 1]
```diff
@@ -265,7 +267,8 @@ class VisitController(
     @PathVariable reference: String,
     @RequestBody @Valid cancelOutcome: OutcomeDto
   ): VisitDto {
-    return visitService.cancelVisit(reference.trim(), cancelOutcome)
+    val principal = SecurityContextHolder.getContext().authentication.principal.toString()
+    return visitService.cancelVisit(reference.trim(), cancelOutcome, principal)
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt:271]: AuthenicationHelper.getUser()

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitControllerLegacy.kt [hunk 2]
```diff
@@ -77,5 +78,8 @@ class VisitControllerLegacy(
     @Schema(description = "reference", example = "v9-d7-ed-7u", required = true)
     @PathVariable reference: String,
     @RequestBody @Valid cancelOutcome: OutcomeDto
-  ): VisitDto = visitService.cancelVisit(reference.trim(), cancelOutcome)
+  ): VisitDto {
+    val principal = SecurityContextHolder.getContext().authentication.principal.toString()
+    return visitService.cancelVisit(reference.trim(), cancelOutcome, principal)
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitControllerLegacy.kt:83]: AuthenicationHelper.getUser()

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt [hunk 3]
```diff
@@ -246,6 +246,10 @@ class VisitService(
         it.outcomeStatus = SUPERSEDED_CANCELLATION
         visitRepository.saveAndFlush(it)
       }
+      visitToBook.createdBy = existingBookedVisit?.createdBy
+      visitToBook.updatedBy = actionedBy
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt:250]: visitToBook.updatedBy = authenicationHelper.getUser()

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt [hunk 4]
```diff
@@ -276,13 +280,11 @@ class VisitService(
...
     val visitEntity = visitRepository.findBookedVisit(reference) ?: throw VisitNotFoundException("Visit $reference not found")
-    val visitCancellationDateAllowed = getAllowedCancellationDate(
-      LocalDateTime.now(), visitCancellationDayLimit = visitCancellationDayLimit.toLong()
-    )
-    validateVisitStartDate(visitEntity, "cancelled", visitCancellationDateAllowed)
+    validateVisitStartDate(visitEntity, "cancelled", getAllowedCancellationDate(visitCancellationDayLimit = visitCancellationDayLimit))
 
     visitEntity.visitStatus = CANCELLED
     visitEntity.outcomeStatus = cancelOutcome.outcomeStatus
+    visitEntity.cancelledBy = actionedBy
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt:287]: = authenicationHelper.getUser()

**Review by AledWEvans — CHANGES_REQUESTED** [review]
move the security context stuff into the service call authenicationHelper.getUser(),
Apart from that look good!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt (hunk 1 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt]: Thanks Aled, have updated based on your comments.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt [hunk 5]
```diff
@@ -217,7 +218,8 @@ class VisitController(
     @Schema(description = "applicationReference", example = "dfs-wjs-eqr", required = true)
     @PathVariable applicationReference: String
   ): VisitDto {
-    return visitService.bookVisit(applicationReference.trim())
+    val principal = SecurityContextHolder.getContext().authentication.principal.toString()
+    return visitService.bookVisit(applicationReference.trim(), actionedBy = principal)
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt]: Thanks, have moved it into a service class as suggested.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/VisitDto.kt [hunk 6]
```diff
@@ -45,6 +45,12 @@ data class VisitDto(
   val visitors: List<VisitorDto> = listOf(),
   @Schema(description = "List of additional support associated with the visit", required = false)
   val visitorSupport: List<VisitorSupportDto> = listOf(),
+  @Schema(description = "Actioned By - NOMIS ID for the user who created the visit", example = "AB12345A", required = false)
+  val createdBy: String?,
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/VisitDto.kt:49]: is this required true

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/Visit.kt [hunk 7]
```diff
@@ -74,6 +74,15 @@ class Visit(
   @OneToMany(fetch = FetchType.LAZY, cascade = [CascadeType.ALL], mappedBy = "visit", orphanRemoval = true)
   var visitNotes: MutableList<VisitNote> = mutableListOf(),
 
+  @Column
+  var createdBy: String? = null,
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/Visit.kt:78]: can this one be null?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/AuthenticationHelperService.kt [hunk 8]
```diff
@@ -0,0 +1,31 @@
+package uk.gov.justice.digital.hmpps.visitscheduler.service
+
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/AuthenticationHelperService.kt:2]: very nice

File: src/main/resources/db/migration/V2_17__add_actioned_by_to_visit.sql [hunk 9]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE visit ADD created_by VARCHAR(30);
```
Comment (AledWEvans) [inline src/main/resources/db/migration/V2_17__add_actioned_by_to_visit.sql:1]: is 30 length the max we will deal with, it's fairy big I suppose

**Review by AledWEvans — COMMENTED** [review]
a few extra comments!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/AuthenticationHelperService.kt (hunk 8 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/AuthenticationHelperService.kt:2]: thanks to your inputs :)

File: src/main/resources/db/migration/V2_17__add_actioned_by_to_visit.sql (hunk 9 above)
Comment (dhirajshettymoj) [reply src/main/resources/db/migration/V2_17__add_actioned_by_to_visit.sql]: thanks, updated to 60 as discussed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/Visit.kt (hunk 7 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/Visit.kt]: thanks, setting this as NOT NULL and as suggested added a default value of NOT_KNOWN

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/VisitDto.kt (hunk 6 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/VisitDto.kt]: updated as suggested.

**Review by AledWEvans — APPROVED** [review]
LGTM

**Pull request merged**